File: ffmpegbar.py
import subprocess as sp
import time
import sys
import json
import os
# import re


def humanReadableTime(seconds):
    hour = int(seconds / 3600)
    minute = int((seconds % 3600) / 60)
    second = int(seconds % 60)
    return f'{hour:02d}:{minute:02d}:{second:02d}'

def humanReadableSize(size, decimal_places=2):
    for unit in ['','K','M','G','T','P','E','Z']:
        if size < 1024.0:
            break
        size /= 1024.0
    return f"{size:.{decimal_places}f} {unit}B"

def progressBar(value, endvalue, start_time, bar_length=20):
    percent = float(value) / endvalue if endvalue else 0
    bar = '[' + '=' * int(round(percent * bar_length) - 1) + '>' + ' ' * (
        bar_length - int(round(percent * bar_length))) + ']'
    time_taken = time.time() - start_time
    eta = (time_taken / value) * (endvalue - value) if value else 0
    finish_at = time.localtime(time.time() + eta)
    return f'{bar} {percent*100:.2f}% {humanReadableTime(time_taken)} / {humanReadableTime(eta)} ({time.strftime("%I:%M:%S %p", finish_at)})'


def getMediaProperties(path):
    proc = sp.Popen(
        "ffprobe -select_streams v:0 -v quiet -print_format json -show_format -show_streams -show_error"
        .split() + [path],
        stdout=sp.PIPE,
        stderr=sp.PIPE)
    out, err = proc.communicate()
    return json.loads(out)["streams"][0]


def parseFfmpegStatus(stdout):
    data = stdout.readline()
    # if the line is empty or not starting with frame=, skip it
    if not data or not data.startswith('frame='):
        return None

    # ==============================================
    # GETTING EVERYTHING FROM FFMPEG PROGRESS OUTPUT
    # ==============================================

    # tidy up the line:
    # - replace multiple spaces with one, remove trailing spaces
    # - split the line with "=" as separator
    # => ['frame', '12 fps', '12 q', '12.3 size', '1234kB time', '12:34:56.78 bitrate', '1234.5kbits/s speed', '1.234x']

    # line = [f.strip() for f in re.sub(' +', ' ', data).split('=')]

    # split each element of the line into a list of two elements if it contains a space
    # => ['frame', '12', 'fps', '12', 'q', '12.3', 'size', '1234kB', 'time', '12:34:56.78', 'bitrate', '1234.5kbits/s', 'speed', '1.234x']

    # for i, f in enumerate(line):
    #     if ' ' in f:
    #         line[i:i + 1] = f.split(' ')
    # data = dict(zip(line[::2], line[1::2]))

    # =====================================
    # JUST GETTING THE FRAME NUMBER AND FPS
    # =====================================

    frame = data.split('frame=')[1].split('fps=')[0].strip()
    fps = data.split('fps=')[1].split('q=')[0].strip()
    return {'frame': frame, 'fps': fps}


def main():

    ffmpeg_params = sys.argv[1:]
    total_frames = int(
        getMediaProperties(ffmpeg_params[ffmpeg_params.index('-i') +
                                         1])['nb_frames'])
    start_time = time.time()
    proc = sp.Popen(['ffmpeg'] + ffmpeg_params,
                    stdout=sp.PIPE,
                    stderr=sp.PIPE,
                    universal_newlines=True)
    while proc.poll() is None:
        data = parseFfmpegStatus(proc.stderr)
        if data is not None:
            print(
                f"{data['frame']} / {total_frames} {data['fps']} {progressBar(int(data['frame']), total_frames, start_time)}",
                end='\r')
        time.sleep(0.1)
    print(
        f"{total_frames} / {total_frames} {data['fps']}fps {progressBar(total_frames, total_frames, start_time)}"
    )

    print()
    output_file = ffmpeg_params[-1]
    data = getMediaProperties(output_file)
    resolution = f"{data['width']}x{data['height']}"
    frame_rate = round(int(data['r_frame_rate'].split('/')[0]) / int(data['r_frame_rate'].split('/')[1]), 2) if '/' in data['r_frame_rate'] else data['r_frame_rate']
    bitrate = f"{round(int(data['bit_rate'])/(10**6), 2)} Mbps"
    out_size = humanReadableSize(os.path.getsize(output_file))
    print(f"==> Output file: {output_file}")
    print(f"- Resolution: {resolution}")
    print(f"- Frame rate: {frame_rate} fps")
    print(f"- Bitrate: {bitrate}")
    print(f"- Size: {out_size}")

if __name__ == '__main__':
    main()

File: ffmpegbar.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"strconv"
	"strings"
	"time"
)

func humanReadableClock(unixTime int64) string {
	return time.Unix(unixTime, 0).Format("15:04:05 PM")
}

func humanReadableTime(unixTime int64) string {
	newTime := time.Unix(unixTime, 0)
	_, offset := newTime.Zone()
	// add the offset to the time
	newTime = newTime.Add(time.Duration(-offset) * time.Second)
	return newTime.Format("15:04:05")
}

func humanReadableSize(size int64, baseUnit string, decimalPlaces int64) string {
	new_size := float64(size)
	var unit string
	for _, unit = range []string{"", "K", "M", "G", "T", "P", "E", "Z", "Y"} {
		if new_size < 1024.0 {
			break
		}
		new_size /= 1024.0
	}

	return fmt.Sprintf("%.*f %s%s", decimalPlaces, new_size, unit, baseUnit)
}

func progressBar(value int64, endvalue int64, startTime int64, barLength int64) string {
	if endvalue == 0 {
		return ""
	}
	if barLength == 0 {
		barLength = 50
	}
	percent := (float64(value) / float64(endvalue)) * 100

	bar_fill := int64((float64(barLength) / 100) * percent)
	bar_empty := barLength - bar_fill
	bar := strings.Repeat("=", int(bar_fill)) + strings.Repeat(" ", int(bar_empty))

	time_taken := time.Now().Unix() - startTime
	time_left := int64(float64(time_taken) / (percent / 100))
	eta := startTime + time_left
	return fmt.Sprintf("%d / %d [%s] %d%% %s / %s (%s) %s", value, endvalue, bar, int(percent), humanReadableTime(time_taken), humanReadableTime(time_left), humanReadableClock(eta), strings.Repeat(" ", 10))

}

type MediaProp struct {
	framerate float64
	frames    int64
	widht     float64
	height    float64
	duration  int64
	bitrate   int64
	size      int64
}

func getMediaProps(path string) (MediaProp, error) {
	// use ffprobe to get media properties: width, height, r_frame_rate, duration, bitrate
	cmd := exec.Command("ffprobe", "-v", "quiet", "-print_format", "json", "-show_format", "-show_streams", path)
	cmd.Stderr = os.Stderr
	// cmd.Stdout = os.Stdout
	output, err := cmd.Output()
	if err != nil {
		return MediaProp{}, err
	}
	var data map[string]interface{}
	if err = json.Unmarshal(output, &data); err != nil {
		return MediaProp{}, err
	}
	stream := data["streams"].([]interface{})[0].(map[string]interface{})
	framerate := stream["r_frame_rate"].(string)
	var framerate_flt float64
	// if framerate contains a slash, evaluate it then assign it to framerate_flt, otherwise just parse it as float
	if strings.Contains(framerate, "/") {
		raw := strings.Split(framerate, "/")
		A, _ := strconv.ParseFloat(raw[0], 64)
		B, _ := strconv.ParseFloat(raw[1], 64)
		framerate_flt = A / B
	} else {
		framerate_flt, _ = strconv.ParseFloat(framerate, 64)
	}
	frames, _ := strconv.ParseInt(stream["nb_frames"].(string), 10, 64)
	width := stream["width"].(float64)
	height := stream["height"].(float64)
	duration, _ := strconv.ParseFloat(stream["duration"].(string), 64)
	bitrate, _ := strconv.ParseInt(stream["bit_rate"].(string), 10, 64)
	size, _ := strconv.ParseInt(data["format"].(map[string]interface{})["size"].(string), 10, 64)
	fmt.Println()
	return MediaProp{framerate_flt, frames, width, height, int64(duration), bitrate, size}, nil
}

func parseFfmpegStatus(stdout string) (int64, float64) {

	if len(stdout) == 0 || !strings.HasPrefix(stdout, "frame=") {
		return 0, 0
	}

	frame := strings.Split(stdout, "frame=")[1]
	frame = strings.Split(frame, "fps=")[0]
	frame = strings.TrimSpace(frame)
	frame_int, _ := strconv.ParseInt(frame, 10, 64)
	fps := strings.Split(stdout, "fps=")[1]
	fps = strings.Split(fps, "q=")[0]
	fps = strings.TrimSpace(fps)
	fps_float, _ := strconv.ParseFloat(fps, 64)

	return frame_int, fps_float
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("No argument")
		os.Exit(1)
	}
	ffmpeg_params := os.Args[1:]

	// Check output file exists
	output_file := ffmpeg_params[len(ffmpeg_params)-1]
	if _, err := os.Stat(output_file); err == nil {
		fmt.Println("Output file already exists")
		os.Exit(1)
	}

	// Check input file exists
	var input_index int
	for i, v := range ffmpeg_params {
		if v == "-i" {
			input_index = i
			break
		}
	}
	input_file := ffmpeg_params[input_index+1]
	if _, err := os.Stat(input_file); err != nil {
		fmt.Println("Input file does not exist")
		os.Exit(1)
	}

	// Get media properties
	media, err := getMediaProps(input_file)
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}

	startTime := time.Now().Unix()

	// Start ffmpeg
	cmd := exec.Command("ffmpeg", ffmpeg_params...)
	ffmpegOutput, _ := cmd.StderrPipe()
	cmd.Start()
	stdout := bufio.NewReader(ffmpegOutput)
	for {
		out, _ := stdout.ReadString('\r')
		if len(out) == 0 {
			break
		}
		frame, fps := parseFfmpegStatus(out)
		fmt.Printf("\r%.2f fps %s", fps, progressBar(frame, media.frames, startTime, 20))
	}
	fmt.Printf("\r%.2ffps %s", 0.0, progressBar(media.frames, media.frames, startTime, 20))
	cmd.Wait()

	// if output file is not /dev/null
	if output_file != "/dev/null" {
		media, err = getMediaProps(output_file)
		if err != nil {
			fmt.Println(err)
			os.Exit(1)
		}
		fmt.Println("\n==> Output file:", ffmpeg_params[len(ffmpeg_params)-1])
		fmt.Println("- Resolution:", media.widht, "x", media.height)
		fmt.Printf("- Framerate: %.2ffps\n", media.framerate)
		fmt.Println("- Duration:", humanReadableTime(media.duration))
		fmt.Println("- Bitrate:", humanReadableSize(media.bitrate, "b", 2))
		fmt.Println("- Size:", humanReadableSize(media.size, "B", 2))
	}
}
